syntax = "proto2";
option cc_generic_services = true;

package ofs.api.master;

enum ErrorCode {
	EC_NONE = 0;
	EC_FILE_NOT_EIXST = 1;
	EC_FILE_EIXST = 2;
	EC_PERMISSION_DENY = 3;
	EC_DIR_NOT_EMPTY = 4;
	EC_AUTHORITY_EXPIRE = 5;
	EC_IS_DIR = 6;
	EC_USER_OR_PASSWORD_ERROR = 7;
	EC_USER_EXPIRE = 8;
	EC_IS_NOT_DIRECTORY = 9;
	EC_ALREADY_EXIST_DELETE_FILE = 10;
	EC_OUT_OF_RANGE = 11;
	EC_BLOCK_NOT_READY = 12;
	EC_BLOCK_MISSING = 13;
}

enum AuthorityType {
	AT_OWNER_READ = 1;
	AT_OWNER_WRITE = 2;
	AT_OWNER_EXECUTE = 4;
	AT_GROUP_READ = 8;
	AT_GROUP_WRITE = 16;
	AT_GROUP_EXECUTE = 32;
	AT_OTHER_READ = 64;
	AT_OTHER_WRITE = 128;
	AT_OTHER_EXECUTE = 256;
}

message LoginReq {
	required string name = 1;
	required string password = 2;
}

message LoginResponse {
	required ErrorCode errCode = 1;
	optional string token = 2;
	optional int32 blockSize = 3;
}

message MakeDirRequest {
	required string token = 1;
	required int32 authority = 2;
	required string directory = 3;
	required string name = 4;
}

message MakeDirResponse {
	required ErrorCode errCode = 1; 
}

message CreateFileRequest {
	required string token = 1;
	required int32 authority = 2;
	required string directory = 3;
	required string name = 4;
}

message CreateFileResponse {
	required ErrorCode errCode = 1; 
}

message ListRequest {
	required string token = 1;
	required string path = 2;
}

message File {
	required string name = 1;
	required string owner = 2;
	required string group = 3;
	required int32 authority = 4;
	required int32 size = 5;
	required int32 blockSize = 6;
	required int64 createTime = 7;
	required int64 updateTime = 8;
	required bool dir = 9;
}

message ListResponse {
	required ErrorCode errCode = 1;
	repeated File files = 2;
}

message RemoveRequest {
	required string token = 1;
	required string path = 2;
}

message RemoveResponse {
	required ErrorCode errCode = 1; 
}

message FileStatusRequest {
	required string token = 1;
	required string path = 2;
}

message FileStatusRespone {
	required ErrorCode errCode = 1; 
	optional File file = 2;
}

message EndPoint {
	required string host = 2;
	required int32 port = 3;
}

message ReadRequest {
	required string token = 1;
	required string path = 2;
	required int32 blockIndex = 3;
}

message ReadResponse {
	required ErrorCode errCode = 1; 
	optional int64 id = 2;
	repeated EndPoint eps = 3;
}

message BlockLease {
	required int64 id = 1;
	required int64 until = 2;
	required int64 version = 3;
	required int64 newVersion = 4;
	required EndPoint ep = 5;
	repeated int32 chunkservers = 6;
	required string key = 7;
}

message WriteRequest {
	required string token = 1;
	required string path = 2;
	required int32 blockIndex = 3;
}

message WriteResponse {
	required ErrorCode errCode = 1; 
	required BlockLease lease = 2;
}

message AppendRequest {
	required string token = 1;
	required string path = 2;
}

message AppendResponse {
	required ErrorCode errCode = 1; 
	required BlockLease lease = 2;
}

service OfsFileService {
	rpc Login (LoginReq) returns (LoginResponse);
	rpc MakeDir (MakeDirRequest) returns (MakeDirResponse);
	rpc Create (CreateFileRequest) returns (CreateFileResponse);
	rpc List (ListRequest) returns (ListResponse);
	rpc Remove (RemoveRequest) returns (RemoveResponse);
	rpc Status (FileStatusRequest) returns (FileStatusRespone);
	rpc Read (ReadRequest) returns (ReadResponse);
	rpc Write (WriteRequest) returns (WriteResponse);
	rpc Append (AppendRequest) returns (AppendResponse);
}
